### ISSUE TITLE
About the training of Generator

### ISSUE THREAD

**Issue by Juanxincaidadao** [brownvc/R3GAN#8]
Thank you for your in-depth and excellent analysis of GAN networks. I have some questions about training the generator. In your code:
During the training of the generator, the gradients of the real samples are also calculated, does this mean that the parameters of the discriminator are updated while training the generator? This seems different from the conventional GAN training procedure.

    def AccumulateGeneratorGradients(self, Noise, RealSamples, Conditions, Scale=1, Preprocessor=lambda x: x):
        FakeSamples = self.Generator(Noise, Conditions)
        RealSamples = RealSamples.detach()
        
        FakeLogits = self.Discriminator(Preprocessor(FakeSamples), Conditions)
        RealLogits = self.Discriminator(Preprocessor(RealSamples), Conditions)
        
        RelativisticLogits = FakeLogits - RealLogits
        AdversarialLoss = nn.functional.softplus(-RelativisticLogits)
        
        (Scale * AdversarialLoss.mean()).backward()
        
        return [x.detach() for x in [AdversarialLoss, RelativisticLogits]]

**Comment by nick-bvc** [comment]
> During the training of the generator, the gradients of the real samples are also calculated

no.

> RealSamples = RealSamples.**_detach()_**

**Comment by nick-bvc** [comment]
The discriminator gradient on real samples is NOT calculated when updating G.
D(real) is a constant, but it cannot be dropped because it is inside the non-linear softplus function, this is what makes RpGAN fundamentally different from WGAN
It's a constant, but it affects the gradient G receives from D.

**Comment by edmondja** [comment]
> The discriminator gradient on real samples is NOT calculated when updating G. D(real) is a constant, but it cannot be dropped because it is inside the non-linear softplus function, this is what makes RpGAN fundamentally different from WGAN It's a constant, but it affects the gradient G receives from D.

I think Juan was more worried about updating the discriminator during D(G(z))'s backprop.

**Comment by Juanxincaidadao** [comment]
My gan model work now with RpGAN loss after adding real sample logits during training G. Thanks.

---Original---
From: ***@***.***&gt;
Date: Fri, Jan 17, 2025 00:01 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [brownvc/R3GAN] About the training of Generator (Issue #8)




  
The discriminator gradient on real samples is NOT calculated when updating G. D(real) is a constant, but it cannot be dropped because it is inside the non-linear softplus function, this is what makes RpGAN fundamentally different from WGAN It's a constant, but it affects the gradient G receives from D.
  
I think Juan was more worried about updating the discriminator during D(G(z))'s backprop.
